### PR TITLE
Fix false sharing in random pool

### PR DESCRIPTION
**Pull request by crtrott**
The pool arrays are accessed by all threads, but the
    elements per thread are rather small and thus share cachelines.

    Making the arrays 2D using the second dimension for padding only,
    solves that problem. I have seen up to 200x improvement for
    20 threads on skylake running the random number example with 100k and 1
    as parameters. Serial performance is slightly reduced, but it only makes
    the "grep a generator slower" by the equivalent of one double load more
    and a few integer ops.

### CODE REVIEW

File: algorithms/src/Kokkos_Random.hpp [hunk 1]
```diff
@@ -902,16 +906,20 @@ class Random_XorShift64_Pool {
...
     state_      = src.state_;
     num_states_ = src.num_states_;
+    padding_    = src.padding_;
     return *this;
   }
 
   void init(uint64_t seed, int num_states) {
     if (seed == 0) seed = uint64_t(1318319);
 
+    padding_    = num_states < 1000 ? 64 : 1;
```
Comment (dalg24) [inline algorithms/src/Kokkos_Random.hpp:916]: No magic numbers.  Add a comment.

**Review by dalg24 — CHANGES_REQUESTED** [review]

File: algorithms/src/Kokkos_Random.hpp (hunk 1 above)
Comment (crtrott) [reply algorithms/src/Kokkos_Random.hpp:917]: I added a comment that this is a magic number :-P

Comment (crtrott) [reply algorithms/src/Kokkos_Random.hpp:917]: seriously I just needed something not too small and not too big 16 or 32 would probably have worked too.

Comment (mhoemmen) [reply algorithms/src/Kokkos_Random.hpp:918]: https://github.com/microsoft/STL/blob/9a9820df1a1d3fa84100e3169ff37fdd4fa41759/stl/inc/new#L40

Comment (mhoemmen) [reply algorithms/src/Kokkos_Random.hpp:918]: That is, `hardware_destructive_interference_size` perhaps?

Comment (dalg24) [reply algorithms/src/Kokkos_Random.hpp:918]: What about the `1000`

**Review by dalg24 — APPROVED** [review]

**Review by DavidPoliakoff — APPROVED** [review]

**Pull request merged**